modules/runtime/nuxeo-runtime-migration/src/main/java/org/nuxeo/runtime/migration/MigrationServiceImpl.java

```diff
@@ -529,4 +533,55 @@ protected void atomic(String id, Consumer<KeyValueStore> consumer) {
         throw new RuntimeException("Cannot lock for write migration: " + id + ", already locked: " + currentLock);
     }
 
+    @Override
+    public Migration getMigration(String id) {
+        MigrationDescriptor descriptor = getDescriptor(XP_CONFIG, id);
+        return new Migration(id, descriptor.getDescription(), descriptor.getDescriptionLabel(), getStatus(id),
+                descriptor.getSteps());
+    }
+
+    @Override
+    public List<Migration> getMigrations() {
+        var descriptors = getDescriptors(XP_CONFIG);
+        return descriptors.stream().map(d -> getMigration(d.getId())).collect(Collectors.toList());
+    }
+
+    @Override
+    public void run(String migrationId) {
+        if(getStatus(migrationId) == null) {
+            probeAndSetState(migrationId);
+        }
+        MigrationDescriptor descriptor = getDescriptor(XP_CONFIG, migrationId);
+        Set<MigrationStepDescriptor> availableSteps = getAvailableSteps(descriptor);
+        if(availableSteps.size() != 1) {
+            throw new IllegalStateException(String.format("Migration %s needs to have exactly one step to run", migrationId));
+        } else {
+            runStep(migrationId, availableSteps.iterator().next().getId());
+        }
+    }
+
+    protected Set<MigrationStepDescriptor> getAvailableSteps(MigrationDescriptor descriptor) {
+        MigrationStatus status = getStatus(descriptor.getId());
+        return descriptor.getSteps()
+                         .entrySet()
+                         .stream()
+                         .filter(e -> e.getValue().getFromState().equals(status.getState()))
+                         .map(Entry::getValue)
+                         .collect(Collectors.toSet());
+    }
```

Comment:
```java
MigrationStepDescriptor step = descriptor.getSteps().values().stream().filter(v -> status.getState().equals(v.getFromState())).collect(Collectors.reducing((a, b) -> null)).orElseThrow(() -> new IllegalStateException(String.format("Migration %s needs to have exactly one step to run", migrationId)));
```

Reply:
Then I would call the method `get{Only | Single}AvailableStep` ?

Reply:
there is actually another case where we need to get all available steps: https://github.com/nuxeo/nuxeo/pull/4027#discussion_r428232972
I think staying in a method that doesn't fail if it finds several steps is more flexible. What do you think ?

Reply:
Don't we want to fail if there'are several steps in this case? I'm not sure the returned `Set` will be useful for the other case.

Reply:
Yes we want to fail.
I changed to a map for the other case, then it can fit the argument.

Otherwise I will have to make 2 methods, 1 to get a single step and fail otherwise and another one to get all available steps